pkg/webhook/util/ip.go

```diff
@@ -0,0 +1,77 @@
+package util
+
+import (
+	"net"
+)
+
+// incrementIP increments the IP address by 1
+func incrementIP(ip net.IP) net.IP {
+	// Convert the IP to a byte slice and increment the last byte
+	ip = ip.To4()
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j]++
+		if ip[j] != 0 {
+			break
+		}
+	}
+	return ip
+}
+
+func GetUsableIPAddressesCount(includeRange string, excludeRange []string) (count int, err error) {
+	var includeRangeList []string
+	includeRangeList = append(includeRangeList, includeRange)
+
+	includeIPAddrList, err := getIPAddressesFromSubnet(includeRangeList, true)
+	if err != nil {
+		return count, err
+	}
+
+	excludeIPAddrList, err := getIPAddressesFromSubnet(excludeRange, false)
+	if err != nil {
+		return count, err
+	}
+
+	for includeIP := range includeIPAddrList {
+		if _, exists := excludeIPAddrList[includeIP]; !exists {
+			count++
+		}
+	}
+
+	return count, nil
+}
+
+func getIPAddressesFromSubnet(ipNetSubnets []string, include bool) (ipAddrList map[string]struct{}, err error) {
+	ipAddrList = make(map[string]struct{})
+
+	for _, ipNetSubnet := range ipNetSubnets {
+		_, network, err := net.ParseCIDR(ipNetSubnet)
+		if err != nil {
+			return ipAddrList, err
+		}
+
+		// Get broadcast address (last address in the subnet)
+		broadcast := getBroadcastAddress(network)
+
+		// Iterate through all the IP addresses in the subnet
+		for ip := network.IP.Mask(network.Mask); network.Contains(ip); incrementIP(ip) {
+			if include && (ip.Equal(network.IP) || ip.Equal(broadcast)) {
```

Comment:
Do we need to check for `broadcast` in the `exclude` range?

Reply:
Sorry, I misunderstood this part. I guess do we really need to compute all the addresses within the include and exclude ranges to get the count? Can we just use the suffix bits to calculate the number of IPs in both lists, make sure they [`overlap`](https://pkg.go.dev/net/netip#Prefix.Overlaps), and count how many are usable?

Reply:
we need the list of IPs to check overlap, so just calculating the number of ips will not be enough.
But instead of adding counting logic, I could use overlap function as you suggested.

Reply:
I want to better understand this. The `netip.Bits()` function can be used to calculate how many IPs in the include and exclude ranges. Based on that, we can calculate how many are usable. Then the `netip.Overlaps()` function just make sure the exclude range is within the include range.

Reply:
its possible , but I feel it becomes even more convoluted as there is no single or straightforward way to do this.
Overlap returns just true or false, we need to again check for each ip to check if its contained in include list.
The prefix length or Bits() could also be checked between include and exclude, but if exclude prefix length is greater we still need what its excluding to calculate the exact number of IPs.

Reply:
Writing a small pseudo code here of what could be possible from my perspective after going through netip package.

func getIPCount(prefixLength int) int {
    return numberofips (from given prefix length)
}

func getFinalIPCount(includePrefix,excludePrefix)  int {
    //Range through excludePrefix.Next()
    
    if includePrefix.Contains(excludePrefix.IP) {
       count++
    }

    return count
}
 
func main() {
    //range this for list of excludePrefix
    // Calculate number of IPs
    includenumIPs := getIPCount(includePrefix.Bits())
    excludenumIPs := getIPCount(excludePrefix.Bits())

    if !includePrefix.Overlaps(excludePrefix) {
        return includenumIPs
    } else {
       return getFinalIPCount()
    }
}